# [[file:../main.org::*IMPORTS][IMPORTS:1]]
import os
import sys
from operator import itemgetter as itget
# IMPORTS:1 ends here

# [[file:../main.org::*FILE CONFIGS][FILE CONFIGS:1]]
configfile: 'config/config.yaml'
# FILE CONFIGS:1 ends here

# [[file:../main.org::*FUNCTIONS][FUNCTIONS:1]]
get_cores_perc = lambda perc: max(1, workflow.cores * perc)
join_path = lambda *args: os.path.join(*args)
# FUNCTIONS:1 ends here

# [[file:../main.org::*GLOBAL VARIABLES][GLOBAL VARIABLES:1]]
output_dirs = config['outputs']
snakefile_path = os.path.dirname(workflow.snakefile)
SAMPLES=list(map(lambda x: os.path.basename(x.path).replace('.fastq',''), filter(lambda x: x.path.endswith('fastq'), os.scandir(config['reads']))))
minia_prefix="outputs/assemblies/minia/minia.k{}.a{}".format(*itget("kmer","abundance")(config['params']['minia']))
# GLOBAL VARIABLES:1 ends here

# [[file:../main.org::*ALL][ALL:1]]
rule all:
    input:
        reads_sets=expand(config['data']+'/merged.{status}.filtering.fastq', status=('before','after')),
        nanoplot_out=expand(itget('before','after')(output_dirs['nanoplot'])),
	minia_assembly=minia_prefix+".contigs.fa",
        miniasm_graphaligner = join_path(config['outputs']['assemblies']['miniasm'], "miniasm.graphaligner.polished.gfa"),
        minia_assembly_gfa_polished=minia_prefix+'.contigs.polished.gfa',
        minipolish_polished=join_path(config['outputs']['assemblies']['miniasm'], "miniasm.minipolish.polished.gfa"),
        figures=expand(join_path('outputs','bandage')+"/{graph}.svg", graph=("miniasm","minipolish","minia"))
        # miniasm_assembly="outputs/assemblies/miniasm/miniasm.polished.gfa",
# ALL:1 ends here

# [[file:../main.org::*Merge reads][Merge reads:1]]
rule merged_reads:
    input:
        expand(config['reads']+'/{sample}.fastq', sample=SAMPLES)
    output:
        config['data']+'/merged.before.filtering.fastq'
    shell:
        "cat {input} > {output}"
# Merge reads:1 ends here

# [[file:../main.org::*NANOPLOT][NANOPLOT:1]]
rule nanoplot:
    input:
        config['data']+"/merged.{status}.filtering.fastq"
    output:
        directory("outputs/nanoplot/{status}_filter")
    threads:
        get_cores_perc(0.5)
    conda:
        "envs/nanoplot_env.yaml"
    shell:
        "NanoPlot -t {threads} --plots dot -o {output} --fastq {input}"
# NANOPLOT:1 ends here

# [[file:../main.org::*FILTER READS][FILTER READS:1]]
rule filter_reads:
    input:
        config['data']+'/merged.before.filtering.fastq'
    output:
        config['data']+'/merged.after.filtering.fastq'
    params:
        **config['params']['filtlong']
    conda:
        "envs/filtlong_env.yaml"
    shell:
        "filtlong --min_length {params.min_length} --keep_percent {params.keep_percent} {input} > {output} "
# FILTER READS:1 ends here

# [[file:../main.org::*MINIA3][MINIA3:1]]
rule minia:
    input:
        config['data']+'/merged.after.filtering.fastq'
    output:
	minia_assembly=minia_prefix+".contigs.fa"
    threads:
        get_cores_perc(0.5)
    params:
        **config['params']['minia'],
        prefix_fasta=minia_prefix
    conda:
        'envs/minia_env.yaml'
    shell:
        "minia -nb-cores {threads} -kmer-size {params.kmer} -abundance-min {params.abundance} -out {params.prefix_fasta} -in {input} && echo {output.minia_assembly}"
# MINIA3:1 ends here

# [[file:../main.org::*FASTA_TO_GFA][FASTA_TO_GFA:1]]
rule minia_fasta_to_gfa:
    input:
	minia_assembly=minia_prefix+".contigs.fa",
        script=join_path(snakefile_path, 'scripts', 'convertToGFA.py'),
    output:
        minia_assembly_gfa=minia_prefix+'.contigs.gfa'
    params:
        **config['params']['minia'],
    conda:
        'envs/minia_env.yaml'
    shell:
        "python {input.script} {input.minia_assembly} {output.minia_assembly_gfa} {params.kmer}"
# FASTA_TO_GFA:1 ends here

# [[file:../main.org::*MINIMAP][MINIMAP:1]]
rule overlap_filtered_reads:
    conda:
        'envs/miniasm_env.yaml'
    input:
        filtlong_reads = config['data']+'/merged.after.filtering.fastq'
    output:
        overlaped_reads = join_path(config['outputs']['assemblies']['miniasm'], "overlap.20Kb.paf")
    threads:
        get_cores_perc(1)
    shell:
        "minimap2 -x ava-ont -t {threads} {input.filtlong_reads} {input.filtlong_reads} > {output.overlaped_reads}"
# MINIMAP:1 ends here

# [[file:../main.org::*MINIASM][MINIASM:1]]
rule miniasm_assembly:
    input:
        filtlong_reads = config['data']+'/merged.after.filtering.fastq',
        overlaped_reads = join_path(config['outputs']['assemblies']['miniasm'], "overlap.20Kb.paf")
    output:
        miniasm_unpolished = join_path(config['outputs']['assemblies']['miniasm'], "miniasm.unpolished.gfa")
    conda:
        'envs/miniasm_env.yaml'
    threads:
        get_cores_perc(0.2)
    shell:
        "miniasm -f {input.filtlong_reads} {input.overlaped_reads} > {output.miniasm_unpolished}"
# MINIASM:1 ends here

# [[file:../main.org::*MINIPOLISH][MINIPOLISH:1]]
rule polishing_miniasm_minipolish:
    input:
        filtlong_reads = config['data']+'/merged.after.filtering.fastq',
        miniasm_unpolished = join_path(config['outputs']['assemblies']['miniasm'], "miniasm.unpolished.gfa")
    output:
        miniasm_minipolish = join_path(config['outputs']['assemblies']['miniasm'], "miniasm.minipolish.gfa")
    threads:
        get_cores_perc(0.5)
    shell:
        "minipolish --threads {threads} {input.filtlong_reads} {input.miniasm_unpolished} > {output.miniasm_minipolish}"
# MINIPOLISH:1 ends here

# [[file:../main.org::*Graphaligner miniasm][Graphaligner miniasm:1]]
rule polishing_graphaligner_miniasm:
    conda:
        'envs/graphaligner_env.yaml'
    input:
        raw_reads=config['data']+'/merged.before.filtering.fastq',
        miniasm_unpolished = join_path(config['outputs']['assemblies']['miniasm'], "miniasm.unpolished.gfa")
    output:
        miniasm_graphaligner = join_path(config['outputs']['assemblies']['miniasm'], "miniasm.graphaligner.polished.gfa"),
        gf_gaf=join_path(config['outputs']['assemblies']['miniasm'], "miniasm.graphaligner.gaf"),
    threads:
        get_cores_perc(0.5)
    params:
        dbtype = "vg",
        seed_minimizer = 15
    shell:
        "GraphAligner -g {input.miniasm_unpolished} -f {input.raw_reads} -x {params.dbtype} --threads {threads} --seeds-minimizer-length {params.seed_minimizer} --seeds-minimizer-windowsize {params.seed_minimizer} -a {output.gf_gaf} --corrected-out {output.miniasm_graphaligner}"
# Graphaligner miniasm:1 ends here

# [[file:../main.org::*Graphaligner minipolish][Graphaligner minipolish:1]]
rule polishing_graphaligner_minipolish:
    conda:
        'envs/graphaligner_env.yaml'
    input:
        raw_reads=config['data']+'/merged.before.filtering.fastq',
        miniasm_minipolish = join_path(config['outputs']['assemblies']['miniasm'], "miniasm.minipolish.gfa")
    output:
        minipolish_gaf=join_path(config['outputs']['assemblies']['miniasm'], "miniasm.minipolish.polished.gaf"),
        minipolish_polished=join_path(config['outputs']['assemblies']['miniasm'], "miniasm.minipolish.polished.gfa")
    threads:
        get_cores_perc(0.5)
    params:
        dbtype = "vg",
        seed_minimizer = 15
    shell:
        "GraphAligner -g {input.miniasm_minipolish} -f {input.raw_reads} -x {params.dbtype} --threads {threads} --seeds-minimizer-length {params.seed_minimizer} --seeds-minimizer-windowsize {params.seed_minimizer} -a {output.minipolish_gaf} --corrected-out {output.minipolish_polished}"
# Graphaligner minipolish:1 ends here

# [[file:../main.org::*Graphaligner MINIA][Graphaligner MINIA:1]]
rule polishing_graphaligner_minia:
    conda:
        'envs/graphaligner_env.yaml'
    input:
        raw_reads=config['data']+'/merged.before.filtering.fastq',
        minia_assembly_gfa=minia_prefix+'.contigs.gfa'
    output:
        minia_gaf=minia_prefix+'.contigs.gaf',
        minia_assembly_gfa_polished=minia_prefix+'.contigs.polished.gfa'
    threads:
        get_cores_perc(0.5)
    params:
        dbtype = "vg",
        seed_minimizer = 15
    shell:
        "GraphAligner -g {input.minia_assembly_gfa} -f {input.raw_reads} -x {params.dbtype} --threads {threads} --seeds-minimizer-length {params.seed_minimizer} --seeds-minimizer-windowsize {params.seed_minimizer} -a {output.minia_gaf} --corrected-out {output.minia_assembly_gfa_polished}"
# Graphaligner MINIA:1 ends here

# [[file:../main.org::*Bandage miniasm][Bandage miniasm:1]]
rule bandage_plot_miniasm:
    input:
        parental_phages = config['parental_phages'],
        minia_graph = join_path(config['outputs']['assemblies']['miniasm'], "miniasm.graphaligner.polished.gfa"),
    output:
        figura=join_path('outputs','bandage')+"/miniasm.svg"
    conda:
        'envs/bandage_env.yaml'
    shell:
        "Bandage image {output.figura} {output.figura} --query {input.parental_phages}"
# Bandage miniasm:1 ends here

# [[file:../main.org::*Bandage minipolish][Bandage minipolish:1]]
rule bandage_plot_minipolish:
    input:
        parental_phages = config['parental_phages'],
        minipolish_polished=join_path(config['outputs']['assemblies']['miniasm'], "miniasm.minipolish.polished.gfa")
    output:
        figura=join_path('outputs','bandage')+"/minipolish.svg"
    conda:
        'envs/bandage_env.yaml'
    shell:
        "Bandage image {output.figura} {output.figura} --query {input.parental_phages}"
# Bandage minipolish:1 ends here

# [[file:../main.org::*Bandage minia][Bandage minia:1]]
rule bandage_plot_minia:
    input:
        parental_phages = config['parental_phages'],
        minia_assembly_gfa_polished=minia_prefix+'.contigs.polished.gfa'
    output:
        figura=join_path('outputs','bandage')+"/minia.svg"
    conda:
        'envs/bandage_env.yaml'
    shell:
        "Bandage image {output.figura} {output.figura} --query {input.parental_phages}"
# Bandage minia:1 ends here
