#+TITLE: Main file of the phage repo
#+AUTHOR: Hugo Ávila (@bioinformagica)
#+LANGUAGE: en-us
#+STARTUP: overview
#+PROPERTY: header-args :dir ~/projects/phage-evo-paper :mkdirp yes :exports none :eval never-export

* ENV setup
** Snakemake
#+BEGIN_SRC shell
# get conda installer
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh

# Run installer and interactively accept and init the conda executable
# Atention !!!: this will change your current shell .rc (.zshrc, .bashrc ...)
bash Miniconda3-latest-Linux-x86_64.sh

# Source the conda changes
source ~/.bashrc

# Set auto activation of conda base environment to false
conda config --set auto_activate_base false

# Add channels
conda config --add channels conda-forge
conda config --add channels bioconda

# Install mamba
conda install -n base -c conda-forge mamba -y

# Install Snakemake
mamba create -c conda-forge -c bioconda -n snakemake snakemake -y

#+END_SRC

#+RESULTS:

* Snakemake workflow template
#+NAME: cb:get-snakemake-template
#+CAPTION: Clone the Snakemake Template
#+BEGIN_SRC shell
git clone https://github.com/snakemake-workflows/snakemake-workflow-template .
rm -rf .git
git init && git commit --allow-empty -m "Initial Commit."
#+END_SRC

* TODO README.md [25%]
#+NAME: cb:README.md
#+CAPTION: README.md
#+BEGIN_SRC markdown :tangle README.md
# Snakemake workflow: `phage-evo-paper`

[![Snakemake](https://img.shields.io/badge/snakemake-≥6.3.0-brightgreen.svg)](https://snakemake.github.io)

A Snakemake workflow for `Phage directed evolution analysis`.
#+END_SRC
** DONE Create README.md
** TODO Create linter action
* TODO Snakefile [12%]
:PROPERTIES:
:COOKIE_DATA: todo recursive
:header-args: :tangle workflow/Snakefile :mkdirp yes :exports none :eval never-export
:END:
** IMPORTS
#+BEGIN_SRC snakemake
import os
import sys
#+END_SRC

** FILE CONFIGS
#+BEGIN_SRC snakemake
configfile: 'config/config.yaml'
#+END_SRC

** GLOBAL VARIABLES
#+BEGIN_SRC snakemake
SAMPLES=list(map(lambda x: os.path.basename(x.path).replace('.fastq',''), filter(lambda x: x.path.endswith('fastq'), os.scandir(config['reads']))))
#+END_SRC

** RULES
*** ALL
#+BEGIN_SRC snakemake
rule all:
    input:
        merged_reads=config['data']+'/all.reads.fastq',
        nanoplot_out=config['nanoplot_out'],
        reads=expand(config['reads']+'/{sample}.fastq', sample=SAMPLES),
        miniasm_assembly=expand("outputs/assemblies/miniasm/{sample}.polished.gfa", sample=SAMPLES)
        # minia_assemby="outputs/assemblies/minia/minia.k{params.kmer}.fasta",
#+END_SRC
*** Merge reads
#+BEGIN_SRC snakemake
rule merged_reads:
    input:
        reads=expand( config['reads']+"/{sample}.fastq", sample=SAMPLES)
    output:
        merged_reads=config['data']+'/all.reads.fastq',
    shell:
        "cat {input.reads} > {output.merged_reads}"
#+END_SRC

*** NANOPLOT
Get some quality check data (mostly read length distribution):
#+BEGIN_SRC snakemake
rule nanoplot:
    conda:
        "envs/nanoplot_env.yaml"
    input:
        merged_reads=config['data']+'/all.reads.fastq',
    output:
        nanoplot_out=directory(config['nanoplot_out'])
    threads:
        workflow.cores
    shell:
        "NanoPlot -t {threads} --plots dot -o {output.nanoplot_out} --fastq {input.merged_reads}"
#+END_SRC

*** MINIA3
Genome assembly with [[https:https://github.com/GATB/minia][minia3]]:
#+BEGIN_SRC snakemake :exports none
# rule minia:
#     conda:
#         'envs/minia_env.yaml'
#     input:
#         merged_reads=config['data']+'/all.reads.fastq',
#     params:
#         kmer=21,
#         abundance=3,
#     threads:
#         workflow.cores
#     output:
#         minia_assemblie=config['minia_out']+"/minia.k{params.kmer}.fasta"
#     shell:
#         "minia -nb-cores {threads} -kmer-size {params.kmer} -abundance-min {params.abundance} -out {output.minia_assemblie} -in {input.merged_reads}"
#+END_SRC

*** MINIASM
#+BEGIN_SRC snakemake
rule miniasm_assembly:
    input:
        config['reads']+'/{sample}.fastq'
    output:
        "outputs/assemblies/miniasm/{sample}.polished.gfa"
    conda:
        'envs/miniasm_env.yaml'
    threads:
        workflow.cores
    shell:
        "miniasm_and_minipolish.sh {input} {threads} > {output}"

#+END_SRC

** TODO Create Rules [1/7]
*** DONE Nanoplot
*** TODO VeChat
*** HOLD Minia3
*** TODO GraphAligner
*** TODO pggb
*** TODO odgi
*** TODO Bonito ???
* CONFIGS
:PROPERTIES:
:COOKIE_DATA: todo recursive
:header-args: :tangle config/config.yaml :mkdirp yes :exports none :eval never-export
:END:
** main file
#+BEGIN_SRC yaml
data: 'data'

# Input
reads: 'data/fastq'

# Outputs
minia_out: 'outputs/minia'
nanoplot_out: 'outputs/nanoplot'
kmergenie_out: 'outputs/kmergenie'
vechat_out: 'outputs/vechat'

# parameters
kmers: [21,33,55,77,99,127]
largest_kmer: 250
#+END_SRC
* ENVS
:PROPERTIES:
:COOKIE_DATA: todo recursive
:header-args: :mkdirp yes :exports none :eval never-export
:END:

#+NAME: get-env-yaml
#+CAPTION: Creates yaml files from conda envs
#+BEGIN_SRC shell :results org replace
declare -a envs=(
    "nanoplot_env"
    "minia_env"
    "kmergenie_env"
)

envs_dir="workflow/envs"
mkdir -p "${envs_dir}"

for env in "${envs[@]}"; do
    fname="${envs_dir}/${env}.yaml"
    [ -f "${fname}" ] && continue
    mamba env export -n "${env}" >"${fname}" 2>/dev/null
done

ls -v1 "${envs_dir}/"*yaml |
    xargs -I'{}' echo "DONE: {}"
#+END_SRC

#+RESULTS: get-env-yaml
#+begin_src org
DONE: workflow/envs/kmergenie_env.yaml
DONE: workflow/envs/minia_env.yaml
DONE: workflow/envs/nanoplot_env.yaml
#+end_src
