#+TITLE: Main file of the phage repo
#+AUTHOR: Hugo Ávila (@bioinformagica)
#+LANGUAGE: en-us
#+STARTUP: overview
#+PROPERTY: header-args :dir ~/projects/phage-evo-paper :mkdirp yes :exports none :eval never-export

* ENV setup
** Snakemake
#+BEGIN_SRC shell
# get conda installer
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh

# Run installer and interactively accept and init the conda executable
# Atention !!!: this will change your current shell .rc (.zshrc, .bashrc ...)
bash Miniconda3-latest-Linux-x86_64.sh

# Source the conda changes
source ~/.bashrc

# Set auto activation of conda base environment to false
conda config --set auto_activate_base false

# Add channels
conda config --add channels conda-forge
conda config --add channels bioconda

# Install mamba
conda install -n base -c conda-forge mamba -y

# Install Snakemake
mamba create -c conda-forge -c bioconda -n snakemake snakemake -y

#+END_SRC

#+RESULTS:

* Snakemake workflow template
#+NAME: cb:get-snakemake-template
#+CAPTION: Clone the Snakemake Template
#+BEGIN_SRC shell
git clone https://github.com/snakemake-workflows/snakemake-workflow-template .
rm -rf .git
git init && git commit --allow-empty -m "Initial Commit."
#+END_SRC

* README.md
#+NAME: cb:README.md
#+CAPTION: README.md
#+BEGIN_SRC markdown :tangle README.md
# Snakemake workflow: `phage-evo-paper`

[![Snakemake](https://img.shields.io/badge/snakemake-≥6.3.0-brightgreen.svg)](https://snakemake.github.io)

A Snakemake workflow for `Phage directed evolution analysis`.
#+END_SRC
** DONE Create README.md
** TODO Create linter action
* Snakefile
:PROPERTIES:
:COOKIE_DATA: todo recursive
:header-args: :tangle workflow/Snakefile :mkdirp yes :exports none :eval never-export :comments link
:END:
** IMPORTS
#+BEGIN_SRC snakemake
import os
import sys
from operator import itemgetter as itget
#+END_SRC

** FILE CONFIGS
#+BEGIN_SRC snakemake
configfile: 'config/config.yaml'
#+END_SRC

** FUNCTIONS
#+BEGIN_SRC snakemake
get_cores_perc = lambda perc: max(1, workflow.cores * perc)
join_path = lambda *args: os.path.join(*args)
get_Kbs = lambda x: int(x/1_000)
#+END_SRC

** GLOBAL VARIABLES
#+BEGIN_SRC snakemake
# output_dirs = config['outputs']
# snakefile_path = os.path.dirname(workflow.snakefile)
# SAMPLES=list(map(lambda x: os.path.basename(x.path).replace('.fastq',''), filter(lambda x: x.path.endswith('fastq'), os.scandir(config['reads']))))

# # Prefixes
# minia_prefix="outputs/assemblies/minia/minia.k{}.a{}".format(*itget("kmer","abundance")(config['params']['minia']))
# pggb_prefix=".p{}.s{}.n{}".format(*itget("map-pct-id", "n-mappings","segment-length")(config['params']['pggb']))
# # Suffixes
# filter_contigs_prefix = minia_prefix + '.min{}Kb.max{}Kb'.format(*iter(map(get_Kbs, itget('min_contig_lenght', 'max_contig_lenght')(config['params']['minia']))))

SAMPLES, = glob_wildcards(join_path(config['data']['reads'], '{sample}.merged.fastq'))
#+END_SRC
** MAIN RULE ALL
#+BEGIN_SRC snakemake
rule all:
    input:
        samples=expand(join_path(config['data']['reads'], '{sample}.merged.fastq'), sample=SAMPLES),
        samples_prefixed=expand(join_path(config['data']['reads'], '{sample}.merged.prefixed.fastq'), sample=SAMPLES)
#+END_SRC

** Include
#+BEGIN_SRC snakemake
include:
  #  'rules/main.smk'
    'rules/parental_genomes_mapping.smk'

#+END_SRC

* Rules
** Main
:PROPERTIES:
:COOKIE_DATA: todo recursive
:header-args: :tangle workflow/rules/main.smk :mkdirp yes :exports none :eval never-export :comments link
:END:
*** ALL
#+BEGIN_SRC snakemake
rule all:
    input:
        minimap2_bam = join_path('results', 'map_reads', 'minimap2_bacterial_and_phage_genomes.bam')
#+END_SRC
*** Merge reads
#+BEGIN_SRC snakemake
rule merged_reads:
    input:
        expand(config['reads']+'/{sample}.fastq', sample=SAMPLES)
    output:
        config['data']+'/merged.before.filtering.fastq'
    shell:
        "cat {input} > {output}"
#+END_SRC

*** NANOPLOT
Get some quality check data (mostly read length distribution):
#+BEGIN_SRC snakemake
rule nanoplot:
    input:
        config['data']+"/merged.{status}.filtering.fastq"
    output:
        directory("outputs/nanoplot/{status}_filter")
    threads:
        get_cores_perc(0.5)
    conda:
        "envs/nanoplot_env.yaml"
    shell:
        "NanoPlot -t {threads} --plots dot -o {output} --fastq {input}"
#+END_SRC

*** FILTER READS
#+BEGIN_SRC snakemake
rule filter_reads:
    input:
        config['data']+'/merged.before.filtering.fastq'
    output:
        config['data']+'/merged.after.filtering.fastq'
    params:
        **config['params']['filtlong']
    conda:
        "envs/filtlong_env.yaml"
    shell:
        "filtlong --min_length {params.min_length} --keep_percent {params.keep_percent} {input} > {output} "
#+END_SRC

*** MINIA3
Genome assembly with [[https:https://github.com/GATB/minia][minia3]]:
#+BEGIN_SRC snakemake
rule minia:
    input:
        config['data']+'/merged.after.filtering.fastq'
    output:
        minia_assembly=minia_prefix+".contigs.fa"
    threads:
        get_cores_perc(0.5)
    params:
        **config['params']['minia'],
        prefix_fasta=minia_prefix
    conda:
        'envs/minia_env.yaml'
    shell:
        "minia -nb-cores {threads} -kmer-size {params.kmer} -abundance-min {params.abundance} -out {params.prefix_fasta} -in {input}"
#+END_SRC

*** FASTA_TO_GFA
#+BEGIN_SRC snakemake
rule minia_fasta_to_gfa:
    input:
        minia_assembly=minia_prefix+".contigs.fa",
        script=join_path(snakefile_path, 'scripts', 'convertToGFA.py'),
    output:
        minia_assembly_gfa=minia_prefix+'.contigs.gfa'
    params:
        **config['params']['minia'],
    conda:
        'envs/minia_env.yaml'
    shell:
        "python {input.script} {input.minia_assembly} {output.minia_assembly_gfa} {params.kmer}"
#+END_SRC

*** Graphaligner MINIA
#+BEGIN_SRC snakemake
rule polishing_graphaligner_minia:
    conda:
        'envs/graphaligner_env.yaml'
    input:
        raw_reads=config['data']+'/merged.before.filtering.fastq',
        minia_assembly_gfa=minia_prefix+'.contigs.gfa'
    output:
        minia_gaf=minia_prefix+'.contigs.gaf',
        minia_assembly_gfa_polished=minia_prefix+'.contigs.polished.fa'
    threads:
        get_cores_perc(1)
    params:
        dbtype = "vg",
        seed_minimizer = 15
    shell:
        "GraphAligner -g {input.minia_assembly_gfa} -f {input.raw_reads} -x {params.dbtype} --threads {threads} --seeds-minimizer-length {params.seed_minimizer} --seeds-minimizer-windowsize {params.seed_minimizer} -a {output.minia_gaf} --corrected-out {output.minia_assembly_gfa_polished}"
#+END_SRC

*** Filter by length
#+BEGIN_SRC snakemake
rule filter_by_length:
    input:
        minia_assembly_gfa_polished=minia_prefix+'.contigs.polished.fa',
        script = join_path(snakefile_path, 'scripts', 'filter_by_length.py')
    output:
        minia_assembly_polished_filtered = minia_prefix + '.contigs.polished' + filter_contigs_prefix + '.fa'
    params:
        **config['params']['minia']
    conda:
        'envs/bio_env.yaml'
    shell:
        "python3 {input.script} {params.min_contig_lenght}  {params.max_contig_lenght} > {output.minia_assembly_polished_filtered}"
#+END_SRC
*** PGGB minia_polished
#+BEGIN_SRC snakemake
rule pggb_minia:
    input:
        minia_assembly_polished_filtered = minia_prefix + '.contigs.polished' + filter_contigs_prefix + '.fa'
    output:
        directory("outputs/pggb/minia"+pggb_prefix)
    params:
        **config['params']['pggb']
    conda:
        'envs/pggb_env.yaml'
    threads:
        get_cores_perc(0.5)
    shell:
        "pggb -i {input.minia_assembly_polished_filtered} -p {params.map-pct-id} -n {params.n-mappings} -s {params.segment-length} -t {threads} -o {output} -m"
#+END_SRC
** Parental Genomes Mapping
:PROPERTIES:
:COOKIE_DATA: todo recursive
:header-args: :tangle workflow/rules/parental_genomes_mapping.smk :mkdirp yes :exports none :eval never-export :comments link
:END:

Objective: Minia assembly > map reads parental / bacteria > remove 100 bacteiral genes > quant chmerical reads

*** Map reads to minia assembly
#+BEGIN_SRC snakemake
rule prefix_fastq:
    input:
        sample=join_path(config['data']['reads'], '{sample}.merged.fastq'),
    output:
        sample_prefixed=join_path(config['data']['reads'], '{sample}.merged.prefixed.fastq')
    threads:
        1
    shell:
        "sed -r '/^@/ s/^@/@{wildcards.sample}#1#/' {input.sample} > {output.sample_prefixed}"
#+END_SRC


* CONFIGS
:PROPERTIES:
:COOKIE_DATA: todo recursive
:header-args: :tangle config/config.yaml :mkdirp yes :exports none :eval never-export :comments link
:END:
#+BEGIN_SRC yaml
# data: 'data'

# # Input
# reads: 'data/fastq'
# parental_phages: 'data/genomes/phage/parental.phage.merged.fasta'

data:
  reads: 'data/fastq'
  genomes:
    ecoli: 'data/genomes/bacteria/E_coli_bl21_noplasmid.fasta'
    bacteira: 'data/genomes/bacteria'
    phages: 'data/genomes/phage'

# Outputs
outputs:
  nanoplot:
    before: 'outputs/nanoplot/before_filter'
    after: 'outputs/nanoplot/after_filter'
  assemblies:
    minia: 'outputs/assemblies/minia'
    miniasm: 'outputs/assemblies/miniasm'

# PARAMETERS

# minia
params:
  minia:
    kmer: 21
    abundance: 7
    min_contig_lenght: 20_000
    max_contig_lenght: 200_000
  filtlong:
    keep_percent: 98
    min_length: 20_000
  pggb:
    map-pct-id: 90
    n-mappings: 100
    segment-length: 1_000
#+END_SRC
* ENVS
:PROPERTIES:
:COOKIE_DATA: todo recursive
:header-args: :mkdirp yes :exports none :eval never-export
:END:

#+NAME: get-env-yaml
#+CAPTION: Creates yaml files from conda envs
#+BEGIN_SRC shell :results org replace
declare -a envs=(
    "nanoplot_env"
    "minia_env"
    "kmergenie_env"
)

envs_dir="workflow/envs"
mkdir -p "${envs_dir}"

for env in "${envs[@]}"; do
    fname="${envs_dir}/${env}.yaml"
    [ -f "${fname}" ] && continue
    mamba env export -n "${env}" >"${fname}" 2>/dev/null
done

ls -v1 "${envs_dir}/"*yaml |
    xargs -I'{}' echo "DONE: {}"
#+END_SRC

#+RESULTS: get-env-yaml
#+begin_src org
DONE: workflow/envs/kmergenie_env.yaml
DONE: workflow/envs/minia_env.yaml
DONE: workflow/envs/nanoplot_env.yaml
#+end_src
